feat: Remove Account Snap Warning (Flask)

owencraston opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

1. What is the reason for the change?
- This change adds an extra warning to users when trying to remove an account snap (https://metamask.github.io/snap-simple-keyring/latest/ for example) AND the user has snap accounts in their wallet
- Snap accounts are not managed/owned by metamask and therefore cannot be recovered in metamask. User's must manage and backup these accounts in the snaps UI.
2. What is the improvement/solution?
- Adds an extra layer of friction when trying to remove an account snap and the user has snap accounts in their wallet
- This is a `FLASK ONLY` feature

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/155

## **Manual testing steps**

1. check this branch
2. open `.js.env`
3. ensure that `export METAMASK_BUILD_TYPE=flask` is set to flask (not main)
4. build the app on this branch
5. create/import a wallet
6. inside the mobile browser, open this URL: https://metamask.github.io/snap-simple-keyring/latest/
7. connect this snap
8. click create account, you can do this as many times as you want.
9. navigate back to the main wallet view
10. NOTICE there should be new accounts in your wallet with the name "Snap Account x"
11. Go to settings
12. navigate to snaps
13. locate the Snap simple keyring settings
14. scroll to the bottom of the page and click the red "Remove" button
15. Notice: new warning should pop up indicating that the user should backup these accounts before removing the snap
16. click continue
17. you should be prompted to type the name of the snap in the text input.
18. YOU SHOULD NOT BE ABLE TO REMOVE THE SNAP WITHOUT TYPING The CORRECT SNAP NAME
19. Once you type the correct snap name, you should be able to remove the snap
20. After clicking remove the snap should be removed.


#### Testing a non account snap
1. check this branch
2. open `.js.env`
3. ensure that `export METAMASK_BUILD_TYPE=flask` is set to flask (not main)
4. build the app on this branch
5. create/import a wallet
6. navigate to settings/snaps
7. click on the pre installed Message signing snap
8. click on the remove button at the bottom
9. `THERE SHOULD NOT BE A WARNING`

#### Testing removing a snap account without snap accounts in the wallet
1. check this branch
2. open `.js.env`
3. ensure that `export METAMASK_BUILD_TYPE=flask` is set to flask (not main)
4. build the app on this branch
5. create/import a wallet
6. inside the mobile browser, open this URL: https://metamask.github.io/snap-simple-keyring/latest/
7. connect this snap. Do not create any accounts this time.
8. Go to settings
9. navigate to snaps
10. locate the Snap simple keyring settings
11. scroll to the bottom of the page and click the red "Remove" button
12. There should be no warning pop up and the snap should be removed

## **Screenshots/Recordings**

### Extension Version

<img width="367" alt="Screenshot 2024-10-01 at 5 38 58 PM" src="https://github.com/user-attachments/assets/024e03f4-bf38-4062-a040-6eb89b1c39c9">
<img width="363" alt="Screenshot 2024-10-01 at 5 39 14 PM" src="https://github.com/user-attachments/assets/918dff2a-ca80-4bee-8889-28042606fe9f">
<img width="361" alt="Screenshot 2024-10-01 at 5 39 20 PM" src="https://github.com/user-attachments/assets/bf27f6bd-b859-49b8-ba84-3d35c399d77c">

### **After**


with snap accounts:

https://github.com/user-attachments/assets/24aa2993-98fa-4d72-bc1f-b7f9c1aeb6ad

<img width="825" alt="Screenshot 2024-10-10 at 3 26 46 PM" src="https://github.com/user-attachments/assets/0026de7c-e657-49ad-b6a6-e085ffc05816">


Without snap accounts:

https://github.com/user-attachments/assets/7cf42c03-dc6e-40fb-b490-fb0eb4d8ba83



## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

owencraston committed:
feat: remove account snap warning

- when a user tries to remove an account snap while they have snap
  accounts in the wallet, we will warn them and prompt them to back up
  these accounts.
- these accounts are not managed by MetaMask so they cannot be recovered

owencraston committed:
tests

owencraston committed:
remove snap if there are no keyring accounts

owencraston committed:
more test

k-g-j commented on app/util/test/accountsControllerTestUtils.ts:
```diff
@@ -39,6 +39,37 @@ export function createMockInternalAccount(
+      },
+      snap: {
+        id: 'npm:@metamask/snap-simple-keyring-snap',
+        name: 'MetaMask Simple Snap Keyring',
```
```suggestion
        name: 'MetaMask Snap Simple Keyring',
```

owencraston replied:
I am going off of the [proposed name of the test](https://github.com/MetaMask/snap-simple-keyring/blob/main/packages/snap/snap.manifest.json#L4C20-L4C48) snap which is `MetaMask Simple Snap Keyring`.

k-g-j commented on app/components/Views/Snaps/SnapSettings/SnapSettings.tsx:
```diff
@@ -42,9 +45,16 @@ const SnapSettings = () => {
+
   // TODO: Replace "any" with type
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   function getPermissionSubjects(state: any) {
```
We could consider using TypeScript's `Record` type instead of `any` for `getPermissionSubjects` and `getPermissions` functions. This would improve type safety and make the code more self-documenting. 

For example:

   ```typescript
   function getPermissionSubjects(state: Record<string, unknown>) {
     return (state.subjects as Record<string, unknown>) || {};
   }

   function getPermissions(state: Record<string, unknown>, origin: string) {
     return (getPermissionSubjects(state)[origin] as Record<string, unknown>)?.permissions;
   }

owencraston replied:
I tried to get proper typing working for this but it did not work since the response from the PermissionController is typed as unknown. Because of this even adding the Record<string, unknown> is going to require a cast of any or unknown since the return object could be {}. In this case you will get the error ```Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'```. Until we can get proper typing on this I say we leave it. At least with the TODO/eslint-disable we can easily find this messy code in an audit of all the any's in the codebase. If you feel strongly against this we can try to get it working.

k-g-j commented on app/components/Views/Snaps/SnapSettings/SnapSettings.tsx:
```diff
@@ -71,52 +81,106 @@ const SnapSettings = () => {
 
+  const isKeyringSnap = Boolean(permissionsFromController?.snap_manageAccounts);
+  
+  useEffect(() => {
```
I think the `useEffect` for `setKeyringAccounts` could be optimized by moving the `getAccountsBySnapId` call outside the effect and using `useMemo` instead. 

```typescript
const keyringAccounts = useMemo(() => {
  if (isKeyringSnap) {
    const addresses = await getAccountsBySnapId(snap.id);
    return Object.values(internalAccounts).filter(
      (internalAccount) =>
        addresses.includes(internalAccount.address.toLowerCase()),
    );
  }
  return [];
}, [snap?.id, internalAccounts, isKeyringSnap]);
```

owencraston replied:
I like this suggestion however in practice it does not work. The issue here is that useMemo doesn't handle async functions in the way we might expect. When you pass an async function to useMemo, it doesn't await the result. Instead, it memoizes the Promise returned by the async function. This results in a return type of Promise<> for the keyringAccounts so we would need to handle the .then of that value all over this file. From what I can see the useeffect + state is the best way to handle this. Let me know if you find a solution. 
<img width="692" alt="Screenshot 2024-10-07 at 8 06 48 PM" src="https://github.com/user-attachments/assets/3b22ef21-e678-4d20-84a1-5aa27a089cec">

gantunesr replied:
An alternative could be wrapping the method inside a useCallback and that invoke it inside the useEffect, something like,

```ts
const fetchIdentities = useCallback(async () => {
  // ...
}, [...]);

useEffect(() => {
  if (isKeyringSnap) {
    fetchIdentities();
  }
}, [fetchIdentities, isKeyringSnap]);
```

k-g-j commented on app/components/Views/Snaps/components/KeyringAccountListItem/KeyringAccountListItem.tsx:
```diff
@@ -0,0 +1,66 @@
+            iconName={IconName.Export}
+            iconColor={IconColor.Primary}
+            size={ButtonIconSizes.Lg}
+            onPress={() => Linking.openURL(snapUrl)}
```
Could the `onPress` handler here for `ButtonIcon` be memoized with `useCallback` to prevent unnecessary re-renders? 

Something along the lines of . . . 

```typescript
  const handlePress = useCallback(() => {
    Linking.openURL(snapUrl);
  }, [snapUrl]);

  return (
    <ButtonIcon
      testID={KEYRING_ACCOUNT_LIST_ITEM_BUTTON}
      iconName={IconName.Export}
      iconColor={IconColor.Primary}
      size={ButtonIconSizes.Lg}
      onPress={handlePress}
    />
  );
}
```

owencraston replied:
yes good catch.

k-g-j commented on app/components/Views/Snaps/KeyringSnapRemovalWarning/test/KeyringSnapRemovalWarning.test.tsx:
```diff
@@ -0,0 +1,317 @@
+      expect(continueButton.props.disabled).toBe(false);
+    });
+  });
+
```
Let's test this edge case too
```suggestion

it('handles empty snap names correctly', () => {
  const { getByTestId } = renderWithProvider(
    <KeyringSnapRemovalWarning
      snap={{ ...mockSnap, manifest: { ...mockSnap.manifest, proposedName: '' } }}
      keyringAccounts={[]}
      onCancel={onCancelMock}
      onClose={onCloseMock}
      onSubmit={onSubmitMock}
    />
  );
  
  const textInput = getByTestId(KEYRING_SNAP_REMOVAL_WARNING_TEXT_INPUT);
  fireEvent.changeText(textInput, '');

  const continueButton = getByTestId(KEYRING_SNAP_REMOVAL_WARNING_CONTINUE);
  expect(continueButton.props.disabled).toBe(true);

  expect(textInput.props.value).toBe('');
  expect(continueButton.props.children[1].props.children).toBe('Remove Snap');
});

```

owencraston replied:
Thank you

k-g-j commented on app/components/Views/Snaps/KeyringSnapRemovalWarning/KeyringSnapRemovalWarning.tsx:
```diff
@@ -0,0 +1,210 @@
+      try {
+        onSubmit();
+      } catch (e) {
+        setError(true);
```
Maybe it would be helpful with debugging if we added some type of more specific error messaging 

something like: 

```typescript
try {
  onSubmit();
} catch (error) {
  if (error instanceof NetworkError) {
    setErrorMessage('Network error occurred. Please try again.');
  } else if (error instanceof ValidationError) {
    setErrorMessage('Validation error: ' + error.message);
  } else {
    setErrorMessage('An unexpected error occurred.');
  }
}
```

owencraston replied:
Based on the extension implementation and designs it seems like we do not  want to render "random" error messages. Instead we render a pre defined error message that can be translated. Instead of setting the error message to be displayed, I will add some logs into this try/catch for better debugging.

<img width="727" alt="Screenshot 2024-10-07 at 8 11 34 PM" src="https://github.com/user-attachments/assets/f31f810a-d380-4a22-a593-e25ce5f196a8">

owencraston committed:
proper dep check in usecallback

k-g-j commented:
Passing initial manually testing! Good job 😄 

https://github.com/user-attachments/assets/904a11fd-14e4-412a-8c76-4f7819ea5b5c

owencraston committed:
fix snap settings test

owencraston committed:
address feedback

owencraston committed:
yarn setup

k-g-j reviewed (dismissed):
Thank you for addressing my comments! I appreciate the changes 😄 

### Manual Testing

**with account snap**

https://github.com/user-attachments/assets/9280188d-aded-4fb9-a564-01d644d85734

**remove a snap account without snap accounts in the wallet**

https://github.com/user-attachments/assets/cdf7c55c-9a3e-4bae-9771-0db653223c7c

gantunesr commented on app/components/Views/Snaps/components/KeyringAccountListItem/KeyringAccountListItem.tsx:
```diff
@@ -0,0 +1,73 @@
+import {
+  KEYRING_ACCOUNT_LIST_ITEM,
+  KEYRING_ACCOUNT_LIST_ITEM_BUTTON,
+} from './KeyringAccountListItem.constants';
+interface KeyringAccountListItemProps {
```
nit

```suggestion
} from './KeyringAccountListItem.constants';

interface KeyringAccountListItemProps {
```

gantunesr commented on app/components/Views/Snaps/components/KeyringAccountListItem/KeyringAccountListItem.tsx:
```diff
@@ -0,0 +1,73 @@
+} from './KeyringAccountListItem.constants';
+interface KeyringAccountListItemProps {
+  account: InternalAccount;
+  snapUrl: string;
```
The name of this attribute confused me on first inspection, I thought it was a link to the snap dapp. What do you think about renaming it to `blockExplorerUrl`?

owencraston replied:
this is better. I made the change in the latest commit.

gantunesr commented on locales/languages/en.json:
```diff
@@ -1035,14 +1035,30 @@
+          "description": "Removing this Snap removes these accounts from MetaMask:",
+          "remove_account_snap_alert_description_1": "Type ",
+          "remove_account_snap_alert_description_2": " to confirm you want to remove this snap:",
+          "banner_title": "Be sure you can access any accounts created by the Snap via private keys before removing them",
```
Is this copy correct? In extension we have: `Be sure you can access any accounts created by this Snap on your own before removing it`

owencraston replied:
@eriknson @AlexJupiter and I changed it based on these discussions in slack. The main reason for this is that we do not use tooltips on mobile (since there is no hover event) so we needed the message to be more clear.

gantunesr replied:
the problem I see with this copy is that not all accounts will have a private key. For example, SCA will be controlled by a smart contract and MPC accounts will rely on private key shards

gantunesr replied:
@owencraston after discussing with @eriknson and @AlexJupiter it was decided to keep the extension copy and open a new issue to improve this UX to provide a better flow for us to understand how to backup their account according to the type.

owencraston replied:
okay I made those changes to the copy in the latest commit. Here is a screenshot

![Simulator Screenshot - iPhone 16 Pro - 2024-10-10 at 13 52 39](https://github.com/user-attachments/assets/dd668cd0-4e0f-4424-8c1f-94fcda99dbad)

owencraston commented on app/core/SnapKeyring/utils/getAccountsBySnapId.ts:
```diff
@@ -0,0 +1,16 @@
+export const getAccountsBySnapId = async (
+  snapId: SnapId,
+) => {
+  const snapKeyring: SnapKeyring = await Engine.getSnapKeyring() as SnapKeyring;
+  return await snapKeyring.getAccountsBySnapId(snapId);
```
I couldn't find an easy way to do this since the response object from the keyring method we are using to get this information is typed as unknown. 

<img width="670" alt="Screenshot 2024-10-09 at 5 30 54 PM" src="https://github.com/user-attachments/assets/25717094-a12c-4a9a-a5ee-8b5822d12d40">

gantunesr replied:
hmm we should improve the return type of that method to not be `unknown[]`. Not a blocker for this PR.

Jonathansoufer reviewed (dismissed):
Agreed w/ Guto's point on comments. Once its addressed, you're good to go. Amazing work as usual!

owencraston committed:
address feedback

owencraston commented:
Addressed comments and it is still working as expected.

![Simulator Screenshot - iPhone 16 Pro - 2024-10-09 at 17 35 40](https://github.com/user-attachments/assets/d70613de-77d2-4df7-8256-6dc9f9f1e390)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-09 at 17 35 44](https://github.com/user-attachments/assets/3c3e58b7-9d20-453e-a489-c3d79aee689a)

Jonathansoufer reviewed (dismissed):
Reapproving after comments addressed

owencraston committed:
update test

owencraston committed:
add missing code fences

Jonathansoufer reviewed (dismissed):
Reapproved.

owencraston committed:
update copy

owencraston committed:
fix test with new copy

owencraston commented:
latest copy version compared to extension

<img width="825" alt="Screenshot 2024-10-10 at 3 26 46 PM" src="https://github.com/user-attachments/assets/5a9f0a52-8add-4904-9118-f458fde5e1e2">

Jonathansoufer reviewed (dismissed):
Re-approved due copy change.

plasmacorral commented:
A few observations that suggest this may not be ready to merge:

1. when more than 3 accounts are present then the title and warning are shifted up and not presented on screen 
<div style="display: inline-block"><img width="150" src="https://github.com/user-attachments/assets/f3aae2f7-7df0-41c3-a373-91d118b5e370"><img width="150" src="https://github.com/user-attachments/assets/e01ae1f7-b8f7-493a-a079-d0bedb3e90e6"><img width="150" src="https://github.com/user-attachments/assets/4feb37d6-83a7-4562-be65-5f382ddde10f"></div>

2. When more than 4 snap accounts are present (will vary based on screen size), the user cannot see or interact with the `Continue` CTA in the remove flow, due to the lack of scrolling support
RECORDING

3. When I disable the snap before removing it, the accounts are not removed.  If I reinstall the snap, I am stuck with multiple accounts in the picker that are not seen or supported with SSK v1.1.6
RECORDING
4. Long press and hold in the accounts picker does not offer the remove account option like we see for imported accounts.  
RECORDING

plasmacorral commented:
Looking at this again before I log off.  1 and 2 above are pretty much the same, but will leave them as-is to avoid creating confusion.

owencraston committed:
remove accounts when removing snap account

owencraston committed:
test removing of accounts

owencraston commented:
@plasmacorral addressed observation 4. Removing the snap now removes the accounts as well. Good catch.


https://github.com/user-attachments/assets/04b259f8-292c-4e86-b9e8-1a701e51276e

owencraston committed:
fix height with many accounts

owencraston commented:
@plasmacorral I addressed the issue with the overflowing accounts list mentioned in observation 1 & 2.

https://github.com/user-attachments/assets/744f74b5-4d16-4601-ae3d-77514ef503d9

owencraston committed:
remove individual snap accounts

owencraston commented:
@mikesposito I have enabled removing individual snap accounts via the three dot menu. I made the copy the same as the copy on extension. 

https://github.com/user-attachments/assets/5350b42d-287a-4f80-9cac-abc409752f25

Copy comparison:

<img width="404" alt="Screenshot 2024-10-17 at 3 47 16 PM" src="https://github.com/user-attachments/assets/80373f1b-60b4-4c3e-91a1-5ca84ec8d667">

![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 15 47 00](https://github.com/user-attachments/assets/2ade1ff2-cd82-4aa0-87d1-3eb560bdbb6a)

owencraston commented:
@mikesposito I tested removing the snap/accounts when the snap is disabled and confirmed that we can remove the snap even when it is disabled. This matches the extension. 

https://github.com/user-attachments/assets/446c871c-f1b0-46d0-99c9-9fcdfdc1fbe4

owencraston committed:
revert .pbxproj

k-g-j reviewed (dismissed):
LGTM! Good work.

preinstalled snap
https://github.com/user-attachments/assets/e7b1266c-d32a-4da5-8904-df3ed546e62b

with account snap
https://github.com/user-attachments/assets/f4d5748f-73e1-4323-b983-4c6639560fa6

without account snap
https://github.com/user-attachments/assets/c000d7d7-2159-450e-8ede-6376e585a3f0

plasmacorral commented:
Took a fresh look at commit bc77790657bf0a5ac29801e2fe5a71cb2cbab6cb on Android 14 with pixel 5a.

Blocking observation remaining-
6. Although the `Remove Snap` title is not pushed off screen now, I am unable to see any more than 2 of the accounts that I am being warned about removing.

See [recording](https://www.loom.com/share/f4ee5568f253411fa47337c330e79c06)

I think that should be resolved before we merge as the user could remove accounts that were not disclosed on-screen in the warning.

Not blocking-
7. The 3 dots to remove the account works well, thank you!  Although I was surprised to see that when I removed the selected account I was always defaulted to Account 1 being selected instead of the next account up.  When I long press and remove an account imported by private key the active account selected is the one immediately above the removed account.  I mention it because it felt odd, but I think may be of no consequence to the user.

Not blocking-
8. The copy being used for removing a single account could still probably use some finesse.  It mentions SRP or private key for this "imported" account.  Since some snap accounts may not have an SRP or a private key (like MPC), we may want to revisit this on both clients.  At minimum we should be consistent in refering to it as a snap account, rather than calling it imported.  That can be handled in a follow-on and should not delay this PR.

Not blocking/Not related to snap accounts-
9. Only now do I appreciate that an imported account cannot be removed from the 3 dots menu in wallet view.  This makes it now the outlier as a removable account that can only be removed through a long press and hold from the accounts picker.  We should at least add the option to remove under 3 dots menu while we consider further consistency with the tap and hold to remove pattern from the account picker.

owencraston committed:
fix scrollview on android

owencraston committed:
do not select first account

owencraston commented:
@plasmacorral Thank you for those observations. I had only tested on iOS and did not notice the issue there. Turns out there were a few extra things I needed to do in order to get the scrolling on android to work inside of an animated view. I also tested on an iPhone mini and the heights/scrolling seem to work well.

Here is a video showing that observation 6 is resolved:

https://github.com/user-attachments/assets/747e176a-0480-492a-998f-681cc7f24a1e

Regarding observation 7. I included that piece of logic to set the selected account to the first account since that was the same logic we used to removing a hardware account from the 3 dot menu. However I do agree this is a weird pattern and differs from the long press logic so I removed that. Now the selected account will be account at index -1. 

Here is a video showing this behaviour.

https://github.com/user-attachments/assets/ea49ad95-5ec0-407d-b523-e0d792641502

Re observation 8. I agree this is not great copy however in order to be consistent with the extension I left it the same. we should probably overhaul this in the future.

Re observation 8: It seems that our removing account logic is all over the place for all sort of account types. Thank you for making that issue so we can resolve it in an isolated change. This will likely need design feedback.

plasmacorral commented:
Based in video posted [here](https://github.com/MetaMask/metamask-mobile/pull/11451#issuecomment-2428120669) it appears that the remaining blocker was resolved.  Moving this to QA pass.

Jonathansoufer commented on app/components/Views/AccountActions/AccountActions.tsx:
```diff
@@ -189,6 +194,49 @@ const AccountActions = () => {
+  /**
+   * Remove the snap account from the keyring
+   */
+  const removeSnapAccount = useCallback(async () => {
```
Is there any chance to a "dirty" state occurs here? Meaning, if something goes wrong while removing the account (through `KeyringController.removeAccount`, how we should handle and inform our users? Should be an "error" state somewhere?

owencraston replied:
This is a valid concern. I dont think we have any designs or patterns for this now. If we look at the other methods for removing accounts they are identical and have no given us issues for a long time. Take removing a hardware account for example.

```
/**
   * Remove the hardware account from the keyring
   * @param keyring - The keyring object
   * @param address - The address to remove
   */
  const removeHardwareAccount = useCallback(async () => {
    if (selectedAddress) {
      await controllers.KeyringController.removeAccount(selectedAddress as Hex);
      await removeAccountsFromPermissions([selectedAddress]);
      trackEvent(MetaMetricsEvents.ACCOUNT_REMOVED, {
        accountType: keyring?.type,
        selectedAddress,
      });
    }
  }, [
    controllers.KeyringController,
    keyring?.type,
    selectedAddress,
    trackEvent,
  ]);
```

If we were take this on I think it should occur in its own isolated change since it would require some design thought.

owencraston replied:
@plasmacorral created an [issue](https://github.com/MetaMask/accounts-planning/issues/653) to improve the consistency of removing accounts, perhaps we could take this on in that issue.

Jonathansoufer commented on app/components/Views/Snaps/SnapSettings/SnapSettings.tsx:
```diff
@@ -70,53 +81,118 @@ const SnapSettings = () => {
+    try {
+      setIsShowingSnapKeyringRemoveWarning(true);
+      removeSnap();
+      setIsShowingSnapKeyringRemoveWarning(false);
```
Not a big deal, but isn't this redundant if you're using `finally` anyway?

Jonathansoufer reviewed (approved):
Left some minor comments, but excellent work overall!

Cal-L reviewed (approved):
Lgtm!
